Cleanup features

CodeSmitty opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Please link to the issue/card here: -->
https://trello.com/c/YWiBYi5C/13-styling-and-ui-cleanup
<!--- Describe your changes in detail -->
I created a toggle function in the recipes controller that is called by the javascript file toggle_favorite. In short this makes it so user can choose to add recipe to favorites from the recipe show page as opposed to using the checkbox on the form. This makes the UI look better and makes for better user experience. 

I also changed the styling on the form page. Mostly tailwind classes.  Added some styling in the application css to manage tom-select styling to make the category input match the rest of the form fields. 

I removed white space and ordered some files. Nothing crazy.
<!--- Why is this change required? What problem does it solve? -->
 This change is to make the UI a little cleaner as well as removing is favorite from the form saving room and making layout more fluid.
<!--- Include screenshots if appropriate -->

## Testing
<!--- Please describe in detail how you tested your changes -->

I did test the labels, I had an issue with the label covering the data making it look blurry. I added a higher translate styling to make the label go higher to fix this issue.

CodeSmitty committed:
initial toggle favorite commit

CodeSmitty committed:
testing heart click

CodeSmitty committed:
testing toggle favorite

CodeSmitty committed:
testing

CodeSmitty committed:
async toggle favorite icon

CodeSmitty committed:
fixed heart layout

CodeSmitty committed:
testing new form layout

CodeSmitty committed:
fixed label positioning when data is present

CodeSmitty committed:
fixed description label

CodeSmitty committed:
Merge branch 'main' into cleanup-features

nfoos commented on app/javascript/controllers/hello_controller.js:
This file can be deleted

nfoos commented on app/views/recipes/new.html.erb:
```diff
@@ -1,5 +1,5 @@
-<div class="mx-auto md:w-2/3 w-full">
-  <h1 class="font-bold text-4xl">New Recipe</h1>
+<div class="mx-auto md:w-2/3 w-full" >
```
whitespace

```suggestion
<div class="mx-auto md:w-2/3 w-full">
```

nfoos replied:
This whitespace is still here

nfoos commented on app/views/recipes/show.html.erb:
```diff
@@ -14,5 +14,7 @@
       <% end %>
     </div>
+
+   
```
whitespace

```suggestion
```

nfoos commented on config/routes.rb:
```diff
@@ -3,4 +3,10 @@
   # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
 
   root 'recipes#index'
+  get 'categories', to: 'categories#create'
```
Not needed, probably came in from merge conflict
```suggestion
```

nfoos commented on config/routes.rb:
```diff
@@ -3,4 +3,10 @@
 
   root 'recipes#index'
+  get 'categories', to: 'categories#create'
+  resources :recipes do
+    member do
+      post :toggle_favorite
+    end
+  end
```
This is fine, but please remove the `resources :recipes` line above

nfoos commented on app/views/layouts/_navigation.html.erb:
```diff
@@ -2,7 +2,7 @@
     <div class="rounded-full bg-white">
         <%= link_to root_path do %>
-          <%= image_tag "Mamezefu-02.png", class: "size-20" %>
+          <%= image_tag "Mamezefu-02.png", class: "size-20 " %>
```
whitespace
```suggestion
          <%= image_tag "Mamezefu-02.png", class: "size-20" %>
```

nfoos commented on app/javascript/controllers/tom_select_controller.js:
```diff
@@ -3,11 +3,14 @@ import TomSelect from 'tom-select'
 export default class extends Controller {
   connect () {
+
+    
     new TomSelect(this.element ,{
         plugins: ['remove_button'],
         create: true,
         createOnBlur: true,
-        persist: false
+        persist: false,
     })
+    
```
Why the changes to this file?

CodeSmitty replied:
Oh, I was trying to see if I could add tailwind classes to tom-select. Which I figured out you can, but not that way and it's more pain than it was worth. This is one of the ways I found, but I didn't look too much into it. https://github.com/orchidjs/tom-select/discussions/160

nfoos replied:
But you left all the whitespace changes. Don't commit the cruft leftover from your experimenting 😛

CodeSmitty replied:
I couldn't remember how the file was originally lol. But yes, you are right.

nfoos commented on app/controllers/recipes_controller.rb:
```diff
@@ -54,6 +54,11 @@ def destroy
   end
 
+  def toggle_favorite
+    @recipe.toggle!(:is_favorite)
```
rubocop
```suggestion
    @recipe.toggle!(:is_favorite) # rubocop:disable Rails/SkipsModelValidations
```

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,69 +11,69 @@
   </div>
-  
+
   <div class="my-5">
     <%= form.label :category_names, "Categories" %>
     <%= form.collection_select :category_names, Category.all, :name, :name, {}, { data: { controller: "tom-select" }, multiple: true } %>
   </div>
 
```
You probably meant to delete this
```suggestion
```

nfoos commented on app/assets/stylesheets/application.css:
```diff
@@ -44,11 +44,36 @@
+  appearance: none !important;
+}
+
+.st-contol:focus > .collection-label > label:focus{
+  display:none;
+}
+
+
```
`.st-contol` is that a typo? Is this needed?

CodeSmitty replied:
It was a typo, but now I can't find the reason that line was there. I removed it for now, it had something to do with the way the category field rendered.

nfoos commented on app/assets/stylesheets/application.css:
```diff
@@ -44,11 +44,36 @@
+}
+
+/* block rounded-t-lg px-2.5 pb-2.5 pt-5 w-full text-sm font-bold text-teal-900  
+border-0 border-b-2 border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer */
```
Is this comment needed?
```suggestion
```

CodeSmitty committed:
fixed review errors

CodeSmitty committed:
Merge branch 'cleanup-features' of https://github.com/CodeSmitty/mame-zefu into cleanup-features

CodeSmitty committed:
removed second collection field

nfoos commented on app/views/recipes/_recipe.html.erb:
```diff
@@ -1,19 +1,28 @@
-        <% recipe.rating.to_i.times do %>
-          <%= render partial: "icons/mini/star", locals: { style: "text-yellow-300" } %>
-        <% end %>
+    <% if recipe.rating.present? %>
```
The diff is all split up so I can't provide a suggestion, but can you move this whole if-block to the same indent level as the `<h1>` tag below?

nfoos commented on app/views/recipes/_recipe.html.erb:
```diff
@@ -1,19 +1,28 @@
-    <% if recipe.is_favorite? %>
-      <%= render partial: "icons/mini/heart", locals: { style: "text-rose-600" } %>
-    <% end %> 
+    <td class="recipe-<%= recipe.id %>">
```
Why a `<td>` tag here? They are typically contained within a `<table>`

nfoos replied:
Also, what's the purpose of setting the `recipe-#` class?

CodeSmitty replied:
I changed the td to a div. This section was a code along from a guide. The td and the class, were the way they wrote it, the idea behind the class was if you needed to modify with document.findElementbyClassName etc. It is not really needed here though. So I removed it. There was no reasoning on my end behind it, it just worked. I tend to forget to modify things I find to my needs.

nfoos commented on app/views/recipes/_recipe.html.erb:
```diff
@@ -1,19 +1,28 @@
+    <td class="recipe-<%= recipe.id %>">
+    <button 
+      type="button" 
+      class="btn btn-xs btn-toggle-favorite text-black" 
```
Is the `btn-toggle-favorite` class doing anything?

CodeSmitty replied:
Nope. Removed it.

nfoos commented on app/views/recipes/_recipe.html.erb:
```diff
@@ -1,19 +1,28 @@
+      data-controller="toggle-favorite" 
+      data-toggle-favorite-url="<%= toggle_favorite_recipe_path(recipe) %>" 
+      data-toggle-favorite-target="button"
+      data-is-favorite="<% recipe.is_favorite %>" >
```
Not sure what this is for, I think it can be removed.
```suggestion
      data-toggle-favorite-target="button" >
```

nfoos replied:
Not sure if you missed this comment, I don't think the `data-is-favorite` is actually doing anything here.

nfoos commented on app/views/recipes/_recipe.html.erb:
```diff
@@ -1,19 +1,28 @@
+      data-toggle-favorite-url="<%= toggle_favorite_recipe_path(recipe) %>" 
+      data-toggle-favorite-target="button"
+      data-is-favorite="<% recipe.is_favorite %>" >
+        <%= render "icons/mini/#{recipe.is_favorite ? 'heart' : 'empty_heart'}" %>
```
I've been trying to pass css classes to the icon partials rather than coding them in the svg. I'm not sure yet if that is a good idea, but I'd like to try staying consistent if possible.
```suggestion
        <%= render partial: "icons/mini/#{recipe.is_favorite ? 'heart' : 'empty_heart'}", locals: { style: "text-rose-600" } %>
```

nfoos replied:
I think you missed this comment

nfoos commented on app/views/icons/mini/_empty_heart.html.erb:
```diff
@@ -1,3 +1,3 @@
-<svg class="h-8 w-8 text-red-500"  fill="none" viewBox="0 0 24 24" stroke="currentColor">
+<svg class="h-5 w-5 text-red-500"  fill="none" viewBox="0 0 24 24" stroke="currentColor">
```
```suggestion
<svg class="h-5 w-5 <%= style if defined?(style) %>"  fill="none" viewBox="0 0 24 24" stroke="currentColor">
```

nfoos replied:
I think you missed this comment

nfoos commented on app/views/icons/mini/_heart.html.erb:
```diff
@@ -1,3 +1,3 @@
-<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="w-5 h-5 <%= style if defined?(style) %>">
+<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="w-5 h-5 <%= style if defined?(style) %> text-rose-600">
```
If the class is passed through the style local, it doesn't have to be set in the svg
```suggestion
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="w-5 h-5 <%= style if defined?(style) %>">
```

nfoos commented on app/javascript/controllers/toggle_favorite_controller.js:
```diff
@@ -0,0 +1,40 @@
+  }
+  
+  updateHeartIcon(isFavorite) {
+    const heart =`<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor" class="w-5 h-5 <%= style if defined?(style) %> text-rose-600">
+                  <path d="m9.653 16.915-.005-.003-.019-.01a20.759 20.759 0 0 1-1.162-.682 22.045 22.045 0 0 1-2.582-1.9C4.045 12.733 2 10.352 2 7.5a4.5 4.5 0 0 1 8-2.828A4.5 4.5 0 0 1 18 7.5c0 2.852-2.044 5.233-3.885 6.82a22.049 22.049 0 0 1-3.744 2.582l-.019.01-.005.003h-.002a.739.739 0 0 1-.69.001l-.002-.001Z" />
+                  </svg>`;
+
+    const heart_outline =`<svg class="h-5 w-5 text-red-500"  fill="none" viewBox="0 0 24 24" stroke="currentColor">
+                          <path stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4.318 6.318a4.5 4.5 0 000 6.364L12 20.364l7.682-7.682a4.5 4.5 0 00-6.364-6.364L12 7.636l-1.318-1.318a4.5 4.5 0 00-6.364 0z"/>
+                          </svg>`
```
I kind of hate that these icons are defined in both the erb and js files. There must be some way they could be defined in just one place.

CodeSmitty replied:
This was one of the areas I had trouble with. I did try to get them to render from javascript from the partial in different ways and I could not really get it to work. In this case, I prefer deleting the partial since they are no longer in use.

nfoos replied:
I'll try to look into this.

nfoos commented on app/javascript/controllers/toggle_favorite_controller.js:
```diff
@@ -0,0 +1,40 @@
+      headers:{
+        'Content-Type': 'application/json'
+      , 'X-CSRF-Token': token
+      }})
```
This file could use some formatting especially in this area.

nfoos commented on app/javascript/controllers/toggle_favorite_controller.js:
```diff
@@ -0,0 +1,40 @@
+
+  async toggleFavorite(event) {
+    const token = document.querySelector('meta[name="csrf-token"]').getAttribute('content');
+    event.preventDefault();
```
I feel like this should be the first line in the method. Although I'm not sure it's technically needed since there is no form, I guess it doesn't hurt.

CodeSmitty replied:
![Screenshot 2024-02-28 085520](https://github.com/CodeSmitty/mame-zefu/assets/51541411/f4c928af-4e17-4343-8bef-60ef7cbde856)

I get an error and it breaks without the crsf-token, so without it, it won't work.

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,69 +11,64 @@
+      <%= form.text_field :name, class: "block rounded-t-lg px-2.5 pb-2.5 pt-5 w-full text-sm font-bold text-teal-900 border-0 border-b-2 border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
+      <%= form.label :name, class: "absolute text-sm text-teal-900 duration-300 transform -translate-y-4 scale-75 top-4 z-10 origin-[0] start-2.5 peer-focus:text-teal-900 peer-placeholder-shown:scale-100 peer-placeholder-shown:translate-y-0 peer-focus:scale-75 peer-focus:-translate-y-4 rtl:peer-focus:translate-x-1/4 rtl:peer-focus:left-auto" %>
+    </div>
+  
```
This should not have been indented

CodeSmitty committed:
fixed some whitespace

CodeSmitty committed:
if block indent

CodeSmitty committed:
removed class from _heart icon

CodeSmitty committed:
moved prevent default

nfoos commented on app/views/recipes/_recipe.html.erb:
```diff
@@ -1,19 +1,27 @@
       <div class="flex flex-row">
         <% recipe.rating.to_i.times do %>
-          <%= render partial: "icons/mini/star", locals: { style: "text-yellow-300" } %>
+            %= render partial: "icons/mini/star", locals: { style: "text-yellow-300" } %>
```
Looks like the erb tag was accidentally broken here
```suggestion
          <%= render partial: "icons/mini/star", locals: { style: "text-yellow-300" } %>
```

CodeSmitty committed:
removed whitespace & data-is-favorite

nfoos committed:
Add helper to render icons

nfoos committed:
Refactor toggle favorite

- use stimulus values to store icon svgs and other attributes
- set initial icon on connect

nfoos committed:
Merge branch 'cleanup-features' into cleanup-features-review

CodeSmitty committed:
Merge pull request #19 from CodeSmitty/cleanup-features-review

Cleanup features review

nfoos commented on app/javascript/controllers/tom_select_controller.js:
```diff
@@ -7,7 +7,7 @@ export default class extends Controller {
         create: true,
         createOnBlur: true,
-        persist: false
+        persist: false,
```
I found this issue https://github.com/orchidjs/tom-select/discussions/38 where the `onItemAdd` callback was suggested and seems to work for clearing the search text after adding a category
```suggestion
        persist: false,
        onItemAdd: function() {
          this.setTextboxValue('');
          this.refreshOptions();
        }
```

nfoos commented:
![image](https://github.com/CodeSmitty/mame-zefu/assets/3910847/23d0593c-6573-4a4d-9160-cb9f43a82c73)

Comments on form styling
- The input field and page background colors are very similar, making it difficult to see where an input starts and ends. I think they need a light border
- The floating labels are visually confusing. This is made worse by the close proximity to other inputs, the lack of borders and the varying sizes when some inputs are filled and others aren't
- The fields are too close together, they need at least `my-10`
- The yield, prep/cook time fields could use a l little space between them
- I think the font weight on the user input should be normal, not bold
- I kind of wish that the text area fields didn't scroll. I don't know if there's an easy way to auto expand the field. That's something that can be researched later
- The placement of the rating field seems off. I think maybe it should be similar to the view page but I don't know if it belongs above the title. Also, the label is kind of awkward. It might not be needed, or it should resemble the labels on the form in regards to placement and font size/weight.

I'd probably rather the form look more along the lines of https://flowbite.com/blocks/application/crud-create-forms/ with some tweaks to the colors and placeholder text

CodeSmitty committed:
Changed layout design and styling. Created auto height onchange for textarea fields

CodeSmitty committed:
fixed textarea sizing, testing fix for floating labels on safari

CodeSmitty committed:
removed whitespace

nfoos commented on app/javascript/controllers/recipe_form_styling_controller.js:
I'd consider a different name for this controller.  Maybe `dynamic_textarea_controller` or just `textarea_controller`

nfoos commented on app/javascript/controllers/recipe_form_styling_controller.js:
```diff
@@ -0,0 +1,18 @@
+import { Controller } from "@hotwired/stimulus";
+
+export default class extends Controller {
+    connect() {
+
+        const tx = document.getElementsByTagName("textarea");
+
+        for (let i = 0; i < tx.length; i++) {
+            tx[i].setAttribute('style', `height: ${tx[i]?.id === 'recipe_description' || tx[i]?.id === 'recipe_notes' ? '70px' : '230px'}`);
+            tx[i].addEventListener("input", OnInput, false);
+        }
+
+        function OnInput() {
+            this.style.height = this?.id === 'recipe_description' || this?.id === 'recipe_notes' ? "70px" : "230px";
+            this.style.height = (this.scrollHeight) + "px";
+        }
+    }
```
Here's a suggestion that addresses the following issues
- Stimulus controllers know what elements they are attached to
- Stimulus provides a way to manage event listeners via actions
- The initial height of the element can be determined roughly using `offsetHeight`
- If you add a few pixels to `scrollHeight`, a scrollbar will not be added to the textarea

```suggestion
  connect() {
    this.initialHeight = this.element.offsetHeight
    this.resizeElement(this.element)
  }

  resize(event) {
    this.resizeElement(event.target)
  }

  resizeElement(element) {
    element.style.height = this.initialHeight + "px";
    element.style.height = element.scrollHeight + 3 + "px";
  }
```

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,71 +10,67 @@
-    <%= form.label :category_names, "Categories" %>
-    <%= form.collection_select :category_names, Category.all, :name, :name, {}, { data: { controller: "tom-select" }, multiple: true } %>
+  <div class="my-10 relative">
+    <%= form.text_area :description, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " "  %>
```
Assuming the controller was renamed and had a `resize` method, this is how it would be attached to this element with the initial size.
```suggestion
    <%= form.text_area :description, data: { controller: "textarea", action: "textarea#resize" }, rows: 2, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " "  %>
```

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,71 +10,67 @@
-    <%= form.collection_select :category_names, Category.all, :name, :name, {}, { data: { controller: "tom-select" }, multiple: true } %>
+  <div class="my-10 relative">
+    <%= form.text_area :description, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " "  %>
+    <%= form.label :description, id: 'labels', class: "absolute text-sm text-teal-900 duration-300 transform -translate-y-4 scale-75 top-2 z-10 origin-[0] bg-gray-50 px-2 peer-focus:bg-gray-50 peer-focus:pb-0 peer-focus:px-2 peer-focus:text-teal-600 peer-placeholder-shown:scale-100 peer-placeholder-shown:-translate-y-1/2 peer-placeholder-shown:top-6 peer-focus:top-2 peer-focus:scale-75 peer-focus ...
```
I don't think this `id` is being used
```suggestion
    <%= form.label :description, class: "absolute text-sm text-teal-900 duration-300 transform -translate-y-4 scale-75 top-2 z-10 origin-[0] bg-gray-50 px-2 peer-focus:bg-gray-50 peer-focus:pb-0 peer-focus:px-2 peer-focus:text-teal-600 peer-placeholder-shown:scale-100 peer-placeholder-shown:-translate-y-1/2 peer-placeholder-shown:top-6 peer-focus:top-2 peer-focus:scale-75 peer-focus:-translate-y-4 rtl:peer-focus:translate-x-1/4 rtl:peer-focus:left-auto start-1" %>
```

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,71 +10,67 @@
-    <%= form.label :description %>
-    <%= form.text_area :description, rows: 4, class: "block shadow rounded-md border border-gray-200 outline-none px-3 py-2 mt-2 w-full" %>
+  <div class="my-10 relative">
+    <%= form.text_area :ingredients, {data: { controller: "recipe-form-styling", target:"recipe-form-styling.ingredient" }, rows: 4, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " "} %>
```
Using suggested controller changes
```suggestion
    <%= form.text_area :ingredients, data: { controller: "textarea", action: "textarea#resize" }, rows: 6, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
```

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,71 +10,67 @@
-    <%= form.label :ingredients %>
-    <%= form.text_area :ingredients, rows: 4, class: "block shadow rounded-md border border-gray-200 outline-none px-3 py-2 mt-2 w-full h-60" %>
+   <div class="my-10 relative">
+    <%= form.text_area :directions, rows: 4, class: "h-60 block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
```
Using suggested controller changes
```suggestion
    <%= form.text_area :directions, data: { controller: "textarea", action: "textarea#resize" }, rows: 6, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
```

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,71 +10,67 @@
-    <%= form.label :directions %>
-    <%= form.text_area :directions, rows: 4, class: "block shadow rounded-md border border-gray-200 outline-none px-3 py-2 mt-2 w-full h-60" %>
+  <div class="my-2 relative">
+    <%= form.text_area :notes, rows: 4, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
```
Using suggested controller changes
```suggestion
    <%= form.text_area :notes, data: { controller: "textarea", action: "textarea#resize" }, rows: 2, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
```

CodeSmitty committed:
changed file name and updated to cleaner  method for textarea resize

CodeSmitty committed:
added whitespace

CodeSmitty committed:
merged main into this branch

nfoos commented on app/views/recipes/_form.html.erb:
```diff
@@ -11,71 +10,67 @@
-    <%= form.label :ingredients %>
-    <%= form.text_area :ingredients, rows: 4, class: "block shadow rounded-md border border-gray-200 outline-none px-3 py-2 mt-2 w-full h-60" %>
+  <div class="my-2 relative">
+    <%= form.text_area :notes, rows: 4, data: { controller: "textarea", action: "textarea#resize" }, rows: 2, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
```
duplicate `rows` key
```suggestion
    <%= form.text_area :notes, data: { controller: "textarea", action: "textarea#resize" }, rows: 2, class: "block rounded-lg px-2.5 pb-2.5 pt-5 w-full text-sm text-gray-900 bg-gray-50  border border-teal-900 appearance-none focus:outline-none focus:ring-0 focus:border-teal-600 peer", placeholder: " " %>
```

nfoos reviewed (changes requested):
Please run `bundle exec erblint -a .` to fix formatting issues in the ERB files

CodeSmitty committed:
removed duplicate rows attribute on :notes text_area

nfoos commented:
Please merge the main branch and run prettier to format javascript and css files

CodeSmitty committed:
ran erblint command and ran prettier ext

CodeSmitty committed:
installed prettier with npm, ran npx prettier --check, did not commit package.json

CodeSmitty committed:
ran -write

nfoos commented on app/javascript/controllers/tom_select_controller.js:
```diff
@@ -4,10 +4,29 @@ import TomSelect from "tom-select"
   connect() {
     new TomSelect(this.element, {
-      plugins: ["remove_button"],
+      plugins: ["remove_button", "input_autogrow", "change_listener"],
```
Is this change_listener plugin needed?

nfoos commented on app/javascript/controllers/tom_select_controller.js:
```diff
@@ -4,10 +4,29 @@ import TomSelect from "tom-select"
+        this.refreshOptions()
+      },
     })
+
+    if (
+      /Android|webOS|iPhone|iPad|iPod|BlackBerry|IEMobile|Opera Mini/i.test(
+        navigator.userAgent,
+      )
+    ) {
+      // true for mobile device
+      const ts = document.getElementsByClassName("ts-control")
+      ts[0].setAttribute("style", "height: 70px !important")
+    } else {
+      const ts = document.getElementsByClassName("ts-control")
+      ts[0].setAttribute("style", "height: 40px !important")
+    }
```
Is the purpose of this to set the initial height of the input when it is empty? It seems to break the input_autogrow plugin. I might rather it look dumb empty than the autogrow not work.

Note: I'm trying to figure out a way to override the tom-select css via tailwind, but not having much luck myself.

CodeSmitty replied:
It's only for mobile, so the first is set to 70px, which fits two rows of categories if needed without the need to auto grow. I'll remove the autogrow and the change listener plugin. I added those to test to see if we could do something like we did with the text_areas but yes. Tom-select is a pain to style. I just forgot to removed them.

nfoos commented on app/views/recipes/new.html.erb:
```diff
@@ -1,9 +1,9 @@
-<div class="mx-auto md:w-2/3 w-full">
-  <h1 class="font-bold text-4xl">New Recipe</h1>
+<div class="mx-auto md:w-4/5 w-full form-container lg:p-40 p-5">
+  <h1 class="font-bold text-2xl lg:text-4xl text-teal-900 text-center text-start">Create your recipe</h1>
```
This width of the containers throughout the app should be consistent. If you are changing the new view, you should update the edit view as well and probably the show view as well.

`lg:p-40` might be a bit excessive
<img width="953" alt="image" src="https://github.com/CodeSmitty/mame-zefu/assets/3910847/fefb8f0d-39ae-473c-8f23-3a112aa5ac0c">

CodeSmitty replied:
I didn't touch those since we were primarily focused on the form. I can add the same padding to the show and edit on this ticket as well if you want. Not a big deal. I found the container, really helped the form align better. I'll test to see if it works well on the show page, the width might need to be different depending on the way the layouts shift.

CodeSmitty committed:
removed ts js for categories input. Fixed padding on edit, show, new

CodeSmitty committed:
styled categories, left placeholder instead

CodeSmitty committed:
Added favicon

nfoos commented on app/views/layouts/application.html.erb:
The `stylesheet_link_tag` for tom-select should be removed from this file

nfoos commented on app/assets/stylesheets/application.tailwind.css:
```diff
@@ -11,3 +12,20 @@
 }
 
 */
+@import "https://cdnjs.cloudflare.com/ajax/libs/tom-select/2.3.1/css/tom-select.min.css";
```
This `@import` is in the file twice. You can also remove the comment above.

nfoos commented on app/views/recipes/edit.html.erb:
```diff
@@ -1,9 +1,9 @@
-<div class="mx-auto md:w-2/3 w-full">
+<div class="mx-auto md:w-4/5 w-full form-container lg:px-20 p-5">
   <h1 class="font-bold text-4xl">Editing recipe</h1>
```
This header should be styled the same as in `new.html.erb`

nfoos commented on app/views/recipes/index.html.erb:
```diff
@@ -2,9 +2,9 @@
+  <div id="recipes" class="w-auto flex-column flex-wrap justify-center align-center">
     <% @recipes.each do |recipe| %>
-      <%= link_to recipe.name, recipe, class: "block font-medium mb-1 text-2xl p-4 text-center" %>
+      <%= link_to recipe.name, recipe, class: "block m-auto w-1/3 font-medium mb-1 text-2xl p-4 text-left" %>
```
Kinda squished on mobile

![image](https://github.com/CodeSmitty/mame-zefu/assets/3910847/abeefeb1-0a54-48ba-b962-cc38879179c6)

CodeSmitty committed:
Fixed mobile/pc layout on index. Removed duplicate imports. Styled Edit Header

nfoos reviewed (approved):
**LGTM** 🌸
